File: m4d/ClientApp/src/models/DanceDatabase/DanceOrder.ts
import { DanceType } from "./DanceType";

export class DanceOrder {
  public dance: DanceType;
  public delta: number;
  public deltaPercent: number;
  public get deltaPercentAbsolute(): number {
    return Math.abs(this.deltaPercent);
  }
  public get deltaMpm(): number {
    return this.delta / this.dance.meter.numerator;
  }

  public static create(dance: DanceType, tempo: number): DanceOrder {
    return new DanceOrder(
      dance,
      dance.tempoRange.calculateDelta(tempo),
      dance.tempoRange.calculateDeltaPercent(tempo),
    );
  }

  private constructor(dance: DanceType, tempoDelta: number, tempoDeltaPercent: number) {
    this.dance = dance;
    this.delta = tempoDelta;
    this.deltaPercent = tempoDeltaPercent;
  }

  public toString(): string {
    const style =
      this.dance.instances.length > 0
        ? this.dance.instances.map((inst) => inst.style).join(", ")
        : "";
    return `${this.dance.name}: Style=(${style}), Delta=(${this.tempoDeltaString})`;
  }

  private get tempoDeltaString(): string {
    const delta = this.deltaMpm;
    return Math.abs(delta) < 0.01
      ? ""
      : delta < 0
        ? `${delta.toFixed(2)}MPM`
        : `+${delta.toFixed(2)}MPM`;
  }
}

File: DanceLib/DanceOrder.cs
using Newtonsoft.Json;

namespace DanceLibrary;

public class DanceOrder
{
    public DanceType Dance { get; }
    public decimal Delta { get; }
    public decimal DeltaPercent { get; set; }
    public decimal DeltaPercentAbsolute => Math.Abs(DeltaPercent);
    public decimal DeltaMpm =>
        new Tempo(Delta).Convert(new TempoType(TempoKind.Mpm, Dance.Meter)).Rate;

    public static DanceOrder Create(DanceType dance, decimal tempo)
    {
        return new DanceOrder(
            dance,
            dance.TempoRange.CalculateDelta(tempo),
            dance.TempoRange.CalculateDeltaPercent(tempo));
    }

    private DanceOrder(DanceType dance, decimal tempoDelta, decimal tempoDeltaPercent)
    {
        Dance = dance;
        Delta = tempoDelta;
        DeltaPercent = tempoDeltaPercent;
    }

    public override string ToString()
    {
        var style = Dance.Instances.Count > 0
            ? string.Join(", ", Dance.Instances.Select(inst => inst.Style))
            : "";
        return $"{Dance.Name}: Style=({style}), Delta=({TempoDeltaString})";
    }
    private string TempoDeltaString
    {
        get
        {
            var delta = DeltaMpm;
            return Math.Abs(delta) < .01M
                ? ""
                : delta < 0 ? $"{delta:F2}MPM" : $"+{delta:F2}MPM";
        }
    }
}
